cmd/zoekt-sourcegraph-indexserver/cleanup.go

```diff
@@ -515,3 +549,68 @@ func removeTombstones(fn string) ([]*zoekt.Repository, error) {
 	}
 	return tombstones, nil
 }
+
+// deleteShards deletes all the shards in indexDir that are associated with
+// the given repoID. If the repository specified by repoID happens to be in a
+// compound shard, the repository is tombstoned instead.
+//
+// deleteShards returns errRepositoryNotFound if the repository specified by repoID
+// isn't present in indexDir.
+//
+// Users must hold the global indexDir lock before calling deleteShards.
+func deleteShards(indexDir string, repoID uint32) error {
+	shardMap := getShards(indexDir)
+
+	shards, ok := shardMap[repoID]
+	if !ok {
+		return errRepositoryNotFound
+	}
+
+	// Ensure that the paths are in reverse sorted order to ensure that Zoekt's repository <-> shard matching logic
+	// works correctly.
+	//
+	// Example: - repoA_v16.00002.zoekt
+	//          - repoA_v16.00001.zoekt
+	//          - repoA_v16.00000.zoekt
+	//
+	// zoekt-indexserver checks whether it has indexed "repoA" by first checking to see if the 0th shard
+	// is present (repoA_v16.00000.zoekt). If it's present, then it gathers all rest of the shards names in ascending order
+	// (...00001.zoekt, ...00002.zoekt). If it's missing, then zoekt assumes that it never indexed "repoA".
+	//
+	// If this function were to crash while deleting repoA, and we only deleted the 0th shard, then shard's 1 & 2 would never
+	// be cleaned up by Zoekt indexserver (since the 0th shard is the only shard that's tested).
+	//
+	// Deleting shards in reverse sorted order (2 -> 1 -> 0) always ensures that we don't leave an inconsistent
+	// state behind even if we crash.
+
+	sort.Slice(shards, func(i, j int) bool {
+		return shards[i].Path > shards[j].Path
+	})
+
+	for _, s := range shards {
+		// Is this repository inside a compound shard? If so, set a tombstone
+		// instead of deleting the shard outright.
+		if zoekt.ShardMergingEnabled() && maybeSetTombstone([]shard{s}, repoID) {
+			shardsLog(indexDir, "tomb", []shard{s})
```

Comment:
Note: Looking at the implementation of shardlog, I'm unsure if it's threadsafe (what happens if there are two writers to the same file)? 

deleteShard's requirement that we hold the global index lock protects us against this, but I wonder if we should make this more explicit (e.g. have a dedicated mutex just for the shardlogger).

Reply:
quoting: https://github.com/natefinch/lumberjack

> Lumberjack assumes that only one process is writing to the output files. Using the same lumberjack configuration from multiple processes on the same machine will result in improper behavior.